src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -192,36 +202,29 @@ public void doOnMessage(Object key, Object payload, KafkaMessageMetadata metadat
 	private Message<Object> toMessage(Object key, Object payload, KafkaMessageMetadata metadata,
 			Acknowledgment acknowledgment) {
 
-		final Map<String, Object> headers = new HashMap<String, Object>();
+		final MessageHeaderAccessor headerAccessor = new MessageHeaderAccessor();
 
-		headers.put(KafkaHeaders.MESSAGE_KEY, key);
-		headers.put(KafkaHeaders.TOPIC, metadata.getPartition().getTopic());
-		headers.put(KafkaHeaders.PARTITION_ID, metadata.getPartition().getId());
-		headers.put(KafkaHeaders.OFFSET, metadata.getOffset());
-		headers.put(KafkaHeaders.NEXT_OFFSET, metadata.getNextOffset());
+		headerAccessor.setHeader(KafkaHeaders.MESSAGE_KEY, key);
+		headerAccessor.setHeader(KafkaHeaders.TOPIC, metadata.getPartition().getTopic());
+		headerAccessor.setHeader(KafkaHeaders.PARTITION_ID, metadata.getPartition().getId());
+		headerAccessor.setHeader(KafkaHeaders.OFFSET, metadata.getOffset());
+		headerAccessor.setHeader(KafkaHeaders.NEXT_OFFSET, metadata.getNextOffset());
 
 		// pre-set the message id header if set to not generate
-		if (!this.generateMessageId) {
-			headers.put(MessageHeaders.ID, MessageHeaders.ID_VALUE_NONE);
-		}
-
-		// pre-set the timestamp header if set to not generate
-		if (!this.generateTimestamp) {
-			headers.put(MessageHeaders.TIMESTAMP, -1L);
-		}
+		 headerAccessor.setLeaveMutable(!(this.generateMessageId || this.generateTimestamp));
 
 		if (!this.autoCommitOffset) {
-			headers.put(KafkaHeaders.ACKNOWLEDGMENT, acknowledgment);
+			headerAccessor.setHeader(KafkaHeaders.ACKNOWLEDGMENT, acknowledgment);
 		}
 
 		if (this.useMessageBuilderFactory) {
 			return getMessageBuilderFactory()
 					.withPayload(payload)
-					.copyHeaders(headers)
+					.copyHeaders(headerAccessor.toMessageHeaders())
 					.build();
 		}
 		else {
-			return new KafkaMessage(payload, headers);
```

Comment:
It makes a copy, in https://github.com/spring-projects/spring-framework/blob/master/spring-messaging/src/main/java/org/springframework/messaging/MessageHeaders.java. It creates a new map from an existing one.. This is called via new GenericMessage.  I remember us going down this path when refactoring last Thursday but will verify.